Distribute rust-docs-json via rustup.

LukeMathWalker opened the pull request:
I am not 100% sure on how to treat `rust-json-docs` in `target_host_combination`. I went along with a similar strategy to the one used for `rust-docs`, but looking for guidance there.

Commit:
Distribute rust-json-docs via rustup.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

Mark-Simulacrum commented on src/tools/build-manifest/src/main.rs:
```diff
@@ -392,6 +393,10 @@ impl Builder {
+
+            // JSON docs are not yet stable, therefore we only include them in the nightly complete
+            // profile for the time being
+            self.extend_profile("complete", &mut manifest.profiles, &["rust-json-docs"]);
```
I wouldn't include them in the complete profile, nightly or not. In general the complete profile isn't really something we need to extend further IMO.

Mark-Simulacrum reviewed (commented):
r=me with complete profile removed

We may also want to remove the check in bootstrap that gates building these on unstable_features(), since we just won't ship the component via the bulid-manifest nightly-only gate.

ehuss commented:
Just to be clear, these are still unstable? Is there a particular reason it doesn't have a preview tag? What is the relationship with #101799? Did that PR only add the ability to build them, but did nothing with the results? Are these built for all targets that have std docs built?

jyn514 commented:
@ehuss that PR built the JSON and added it RLO, but I didn't realize changing the manifest it's a separate step, so currently is only possible to download the json manually, not through rustup.

Using `-preview` sounds correct to me, we only distribute this component on nightly currently.

Yes, this is for all targets that have docs built. Yes, the json format is still unstable (similar to save-analysis, but we plan to stabilize it in the near future).

LukeMathWalker commented:
Based on the comments above:
- I've marked `rust-json-docs` as a preview component;
- Removed `rust-json-docs` from the `complete` profile;
- Removed the nightly logic from dist, relying on build-manifest to ensure that it doesn't get shipped on stable/beta.

Let me know if you have further comments @Mark-Simulacrum!

Mark-Simulacrum commented on src/tools/build-manifest/src/main.rs:
```diff
@@ -318,6 +318,7 @@ impl Builder {
         package!("rust-mingw", MINGW);
         package!("rust-std", TARGETS);
         self.package("rust-docs", &mut manifest.pkg, HOSTS, DOCS_FALLBACK);
+        self.package("rust-json-docs-preview", &mut manifest.pkg, HOSTS, DOCS_FALLBACK);
```
Hm, so looking at this, can we rename the component to rust-docs-json-preview? I think keeping that common prefix will help with sorting the files etc and feels a little cleaner to me. I think we call them "JSON docs" in some places, so if you disagree I don't feel strongly at all about this and we can keep the current name (renaming it is more of a pain down the line).

LukeMathWalker replied:
"JSON docs" feels more natural than "docs JSON", but I see your point. I don't feel strongly about either - @jyn514?

jyn514 replied:
The other 3-compenent name is `llvm-tools-preview`. The other two doc components are `rust-docs` and `rustc-docs`. `rustc-docs-json` and `rustc-json-docs` both seem like reasonable extensions to me, so I don't think there's an indicator there. But going from most specific seems good to me; `rust-json-docs` kind of sounds like documentation about save-analysis or something. So I have a slight preference for `rust-docs-json`.

LukeMathWalker replied:
I've renamed everything to `rust-docs-json` and squashed all commits 👍🏻

Mark-Simulacrum commented:
@bors r+ rollup

bors commented:
:pushpin: Commit e572d513bd2ca1743d141ccee91563d6faf2e19f has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum e572d513bd2ca1743d141ccee91563d6faf2e19f -->
<!-- homu: {"type":"Approved","sha":"e572d513bd2ca1743d141ccee91563d6faf2e19f","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
